Deno test fails with inscrutable error when `assertSnapshot` is used in update mode but the `--allow-write` flag is missing

**Describe the bug**

Deno test fails with inscrutable error when `assertSnapshot` is used in update mode but the `--allow-write` flag is missing. This is hard to debug without knowledge of the inner workings of `assertSnapshot`.

**Steps to Reproduce**

1. In a new folder, create the file `main_test.ts` with the following contents:

```ts
// main_test.ts
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("should match the snapshot", async (context) => {
  await assertSnapshot(context, "hello world!");
});
```

2. Run `deno test --allow-read -- -u` and observe the output:

```sh
$ deno test --allow-read -- -u
running 1 test from ./main_test.ts
should match the snapshot ... ok (2ms)
Uncaught error from ./main_test.ts FAILED

 ERRORS 

./main_test.ts (uncaught error)
error: null
This error was not caught from a test and caused the test runner to fail on the referenced module.
It most likely originated from a dangling promise, event/timeout handler or top-level code.

 FAILURES 

./main_test.ts (uncaught error)

FAILED | 1 passed | 1 failed (5ms)

error: Test failed
```

**Expected behavior**

<!-- A clear and concise description of what you expected to happen. -->

Ideally, a sensible error would be logged, such as:

```sh
$ deno test --allow-read -- -u
running 1 test from ./main_test.ts
should match the snapshot ... FAILED (0ms)

 ERRORS 

should match the snapshot => ./main_test.ts:4:6
error: AssertionError: The --allow-write flag must be passed when updating snapshots.
    throw new AssertionError(msg);
          ^
    at assert (https://jsr.io/@std/assert/1.0.0-rc.2/assert.ts:21:11)
    at ...

 FAILURES 

should match the snapshot => ./main_test.ts:4:6

FAILED | 0 passed | 1 failed (1ms)

error: Test failed
```

**Environment**

<!-- please complete the following information -->

- OS: [e.g. Ubuntu 20.04, MacOS 11] MacOS Ventrura 13.6.6
- deno version:
```
deno 1.44.4 (release, aarch64-apple-darwin)
v8 12.6.228.9
typescript 5.4.5
```
- std version: 0.225.2

--- a/testing/snapshot.ts
+++ b/testing/snapshot.ts
@@ -438,8 +438,17 @@ class AssertSnapshotContext {
    * This method can safely be called more than once and will only register the teardown
    * function once in a context.
    */
-  registerTeardown() {
+  async registerTeardown() {
     if (!this.#teardownRegistered) {
+      const permission = await Deno.permissions.query({
+        name: "write",
+        path: this.#snapshotFileUrl,
+      });
+      if (permission.state !== "granted") {
+        throw new Deno.errors.PermissionDenied(
+          `Missing write access to snapshot file (${this.#snapshotFileUrl}). This is required because assertSnapshot was called in update mode. Please pass the --allow-write flag.`,
+        );
+      }
       globalThis.addEventListener("unload", this.#teardown);
       this.#teardownRegistered = true;
     }
@@ -582,7 +591,7 @@ export async function assertSnapshot(
   const _serialize = options.serializer || serialize;
   const _actual = _serialize(actual);
   if (getIsUpdate(options)) {
-    assertSnapshotContext.registerTeardown();
+    await assertSnapshotContext.registerTeardown();
     if (!equal(_actual, snapshot)) {
       assertSnapshotContext.updateSnapshot(name, _actual);
     }

--- a/testing/snapshot_test.ts
+++ b/testing/snapshot_test.ts
@@ -863,10 +863,12 @@ Deno.test(
 
     await assertSnapshot(
       t,
-      formatTestOutput(errors).replaceAll(tempDir, "<tempDir>").replace(
-        /Snapshot Test => .+\n/g,
-        "",
-      ),
+      formatTestOutput(errors)
+        .replace(/\s+at .+\n/g, "")
+        .replace(
+          /\nSnapshot Test => .+\n/g,
+          "",
+        ),
     );
     assert(!output.success, "The test should fail");
   }),

--- a/testing/snapshot_test.ts
+++ b/testing/snapshot_test.ts
@@ -9,6 +9,7 @@ import {
   fail,
 } from "@std/assert";
 import { assertSnapshot, createAssertSnapshot, serialize } from "./snapshot.ts";
+import { ensureDir } from "../fs/ensure_dir.ts";
 
 const SNAPSHOT_MODULE_URL = toFileUrl(join(
   dirname(fromFileUrl(import.meta.url)),
@@ -20,7 +21,7 @@ function formatTestOutput(string: string) {
   return stripAnsiCode(string).replace(/([0-9])+m?s/g, "--ms").replace(
     /(?<=running ([0-9])+ test(s)? from )(.*)(?=test.ts)/g,
     "<tempDir>/",
-  );
+  ).replace(/\(file:\/\/.+\)/g, "(file://<path>)");
 }
 
 function formatTestError(string: string) {
@@ -814,3 +815,56 @@ Deno.test("createAssertSnapshot()", async (t) => {
     );
   });
 });
+
+// Regression test for https://github.com/denoland/deno_std/issues/5155
+// Asserting snapshots in update mode without required write permissions
+Deno.test(
+  "assertSnapshot() - regression #5155",
+  testFnWithTempDir(async (t, tempDir) => {
+    const tempTestFileName = "test.ts";
+    const tempTestFilePath = join(tempDir, tempTestFileName);
+    const tempSnapshotFileName = `${tempTestFileName}.snap`;
+    const tempSnapshotDir = join(
+      tempDir,
+      "__snapshots__",
+    );
+    const tempSnapshotFilePath = join(tempSnapshotDir, tempSnapshotFileName);
+    await ensureDir(tempSnapshotDir);
+    await Deno.writeTextFile(
+      tempSnapshotFilePath,
+      [
+        "export const snapshot = {};",
+        "\nsnapshot[`Snapshot Test 1`] = \`42\`;",
+      ].join("\n"),
+    );
+    await Deno.writeTextFile(
+      tempTestFilePath,
+      `
+      import { assertSnapshot } from "${SNAPSHOT_MODULE_URL}";
+
+      Deno.test("Snapshot Test", async (t) => {
+        await assertSnapshot(t, 42);
+      });
+    `,
+    );
+
+    const command = new Deno.Command(Deno.execPath(), {
+      args: [
+        "test",
+        "--no-lock",
+        "--allow-read",
+        tempTestFilePath,
+        "--",
+        "-u",
+      ],
+    });
+    const output = await command.output();
+    const errors = new TextDecoder().decode(output.stdout);
+
+    await assertSnapshot(
+      t,
+      formatTestOutput(errors).replaceAll(tempDir, "<tempDir>"),
+    );
+    assert(!output.success, "The test should fail");
+  }),
+);

--- a/testing/snapshot_test.ts
+++ b/testing/snapshot_test.ts
@@ -863,7 +863,10 @@ Deno.test(
 
     await assertSnapshot(
       t,
-      formatTestOutput(errors).replaceAll(tempDir, "<tempDir>"),
+      formatTestOutput(errors).replaceAll(tempDir, "<tempDir>").replace(
+        /Snapshot Test => .+\n/g,
+        "",
+      ),
     );
     assert(!output.success, "The test should fail");
   }),

